### PR TITLE
String marshallers for UTF-16 and UTF-8

### PR DESCRIPTION
**Pull request by elinor-fung**
- String marshallers for UTF-16 and UTF-8
- Require explicit marshalling information (`CharSet` or `MarshalAs`) for `char` or `string`
  - See updated compatibility doc
- Add `CharEncoding.Ansi`
  - Not supported for `char`, support for `string` will be added in another change
- Add and update tests

There is currently a good amount of duplication between `Utf16StringMarshaller` and `Utf8StringMarshaller`. I'm planning on doing some refactoring to pull out common bits when I deal with `CharSet.Ansi` and `CharSet.Auto` in a subsequent change.

cc @AaronRobinsonMSFT @jkoritzinsky 

UTF-16:
```C#
public static partial string Method(string p, in string pIn, ref string pRef, out string pOut)
{
    unsafe
    {
        pOut = default;
        string __retVal;
        //
        // Setup
        //
        ushort *__retVal_gen_native;
        ushort *__pIn_gen_native;
        ushort *__pRef_gen_native;
        ushort *__pOut_gen_native;
        //
        // Marshal
        //
        bool pIn__usedCoTaskMem = false;
        if (pIn == null)
        {
            __pIn_gen_native = null;
        }
        else
        {
            int pIn__byteLen = (pIn.Length + 1) * sizeof(ushort);
            if (pIn__byteLen > 520)
            {
                __pIn_gen_native = (ushort *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni(pIn);
                pIn__usedCoTaskMem = true;
            }
            else
            {
                ushort *pIn__stackalloc = stackalloc ushort[pIn.Length + 1];
                ((System.ReadOnlySpan<char>)pIn).CopyTo(new System.Span<char>(pIn__stackalloc, pIn.Length + 1));
                __pIn_gen_native = pIn__stackalloc;
            }
        }

        __pRef_gen_native = (ushort *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni(pRef);
        //
        // Invoke
        //
        fixed (char *__p_gen_native = p)
        {
            __retVal_gen_native = Method__PInvoke__((ushort *)__p_gen_native, &__pIn_gen_native, &__pRef_gen_native, &__pOut_gen_native);
        }

        //
        // Unmarshal
        //
        __retVal = __retVal_gen_native == null ? null : new string ((char *)__retVal_gen_native);
        pRef = __pRef_gen_native == null ? null : new string ((char *)__pRef_gen_native);
        pOut = __pOut_gen_native == null ? null : new string ((char *)__pOut_gen_native);
        //
        // Cleanup
        //
        if (pIn__usedCoTaskMem)
        {
            System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__pIn_gen_native);
        }

        System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__pRef_gen_native);
        System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__pOut_gen_native);
        return __retVal;
    }

    [System.Runtime.InteropServices.DllImportAttribute("DoesNotExist", CharSet = System.Runtime.InteropServices.CharSet.Unicode, EntryPoint = "Method")]
    extern private static unsafe ushort *Method__PInvoke__(ushort *p, ushort **pIn, ushort **pRef, ushort **pOut);
}
```

UTF-8:
```C#
public static partial string Method(string p, in string pIn, ref string pRef, out string pOut)
{
    unsafe
    {
        pOut = default;
        string __retVal;
        //
        // Setup
        //
        byte *__retVal_gen_native;
        byte *__p_gen_native;
        byte *__pIn_gen_native;
        byte *__pRef_gen_native;
        byte *__pOut_gen_native;
        //
        // Marshal
        //
        bool p__usedCoTaskMem = false;
        if (p == null)
        {
            __p_gen_native = null;
        }
        else
        {
            int p__byteLen = (p.Length + 1) * 3 + 1;
            if (p__byteLen > 260)
            {
                __p_gen_native = (byte *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUTF8(p);
                p__usedCoTaskMem = true;
            }
            else
            {
                byte *p__stackalloc = stackalloc byte[p__byteLen];
                p__byteLen = System.Text.Encoding.UTF8.GetBytes(p, new System.Span<byte>(p__stackalloc, p__byteLen));
                p__stackalloc[p__byteLen] = 0;
                __p_gen_native = (byte *)p__stackalloc;
            }
        }

        bool pIn__usedCoTaskMem = false;
        if (pIn == null)
        {
            __pIn_gen_native = null;
        }
        else
        {
            int pIn__byteLen = (pIn.Length + 1) * 3 + 1;
            if (pIn__byteLen > 260)
            {
                __pIn_gen_native = (byte *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUTF8(pIn);
                pIn__usedCoTaskMem = true;
            }
            else
            {
                byte *pIn__stackalloc = stackalloc byte[pIn__byteLen];
                pIn__byteLen = System.Text.Encoding.UTF8.GetBytes(pIn, new System.Span<byte>(pIn__stackalloc, pIn__byteLen));
                pIn__stackalloc[pIn__byteLen] = 0;
                __pIn_gen_native = (byte *)pIn__stackalloc;
            }
        }

        __pRef_gen_native = (byte *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUTF8(pRef);
        //
        // Invoke
        //
        __retVal_gen_native = Method__PInvoke__(__p_gen_native, &__pIn_gen_native, &__pRef_gen_native, &__pOut_gen_native);
        //
        // Unmarshal
        //
        __retVal = System.Runtime.InteropServices.Marshal.PtrToStringUTF8((System.IntPtr)__retVal_gen_native);
        pRef = System.Runtime.InteropServices.Marshal.PtrToStringUTF8((System.IntPtr)__pRef_gen_native);
        pOut = System.Runtime.InteropServices.Marshal.PtrToStringUTF8((System.IntPtr)__pOut_gen_native);
        //
        // Cleanup
        //
        if (p__usedCoTaskMem)
        {
            System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__p_gen_native);
        }

        if (pIn__usedCoTaskMem)
        {
            System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__pIn_gen_native);
        }

        System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__pRef_gen_native);
        System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__pOut_gen_native);
        return __retVal;
    }
}

[System.Runtime.InteropServices.DllImportAttribute("DoesNotExist", EntryPoint = "Method")]
extern private static unsafe byte *Method__PInvoke__(byte *p, byte **pIn, byte **pRef, byte **pOut);
```

### CODE REVIEW

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                            string stackAllocPtrIdentifier = $"{managedIdentifier}__stackalloc";
+                            string fixedCharIdentifier = $"{managedIdentifier}__fixedChar";
+                            string byteLenIdentifier = $"{managedIdentifier}__byteLen";
+
+                            // Code block for stackalloc if string is below threshold size
+                            var marshalOnStack = Block(
+                                // byte* <stackAllocPtr> = stackalloc byte[<byteLen>];
+                                LocalDeclarationStatement(
+                                    VariableDeclaration(
+                                        PointerType(PredefinedType(Token(SyntaxKind.ByteKeyword))),
```
Comment (jkoritzinsky) [inline DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:96]: I think if we use the `ReadOnlySpan<char>, Span<byte>` override for `GetBytes`, the code will be easier to read and won't require us to pin the string. We'll have to do some conversion for the argument from a byref to a pointer, but that's pretty simple to do at that point.

(You can assign a stackalloc to a `Span<T>` without a cast. Roslyn generates the constructor call for you with the correct length).

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                                            NativeType,
+                                            IdentifierName(stackAllocPtrIdentifier)))));
+
+                            // if (<managed> == null)
+                            // {
+                            //     <native> = null;
+                            // }
+                            // else
+                            // {
+                            //     int <byteLen> = (<managed>.Length + 1) * 3 + 1;
```
Comment (jkoritzinsky) [inline DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:160]: Maybe use `Encoding.UTF8.GetMaxByteCount` or a comment explaining this calculation?

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+                            //     {
+                            //         <nativeIdentifier> = (ushort*)Marshal.StringToCoTaskMemUni(<managedIdentifier>);
+                            //         <usedCoTaskMemIdentifier> = true;
+                            //     }
+                            //     else
+                            //     {
+                            //         ushort* <stackAllocIdentifier> = stackalloc ushort[<managedIdentifier>.Length + 1];
+                            //         fixed (char* <fixedCharIdentifier> = <managedIdentifier>)
+                            //         {
+                            //             Buffer.MemoryCopy(<fixedCharIdentifier>, <stackAllocIdentifier>, <byteLenIdentifier>, <byteLenIdentifier>);
```
Comment (jkoritzinsky) [inline DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs:194]: Instead of assigning the stackalloc to a pointer, pinning the string and using `Buffer.MemoryCopy`, it might be worth assigning the stackalloc to a `Span<char>` or `Span<ushort>` and using `ReadOnlySpan<T>.CopyTo` instead. That way we don't need to pin the string.

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                                        IdentifierName(stackAllocPtrIdentifier))));
+
+                            // if (<managedIdentifier> == null)
+                            // {
+                            //     <nativeIdentifier> = null;
+                            // }
+                            // else
+                            // {
+                            //     int <byteLenIdentifier> = (<managedIdentifier>.Length + 1) * sizeof(ushort);
+                            //     if (<byteLenIdentifier> > <StackAllocBytesThreshold>)
```
Comment (jkoritzinsky) [inline DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs:184]: Maybe we want to abstract out the conditional stackalloc concept to a base class? We already do this twice and arrays will do this as well.

**Review by jkoritzinsky — COMMENTED** [review]
Just a few comments around some possible refactoring and maybe using Span in the generated code.

File: DllImportGenerator/DllImportGenerator.IntegrationTests/CharacterTests.cs [hunk 5]
```diff
@@ -7,12 +7,6 @@ namespace DllImportGenerator.IntegrationTests
 {
     partial class NativeExportsNE
     {
-        [GeneratedDllImport(nameof(NativeExportsNE), EntryPoint = "unicode_return_as_uint")]
-        public static partial uint ReturnDefaultAsUInt(char input);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator.IntegrationTests/CharacterTests.cs:11]: Do we not want these tests?

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs (hunk 2 above)
Comment (elinor-fung) [reply DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:160]: Huh, I thought I had a comment. Must have lost it during some moving around of things...
I had kept it as the straight calculation to match what built-in and avoid the extra call and checks that `GetMaxByteCount` does.

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.Interop
+{
+    internal class Utf16StringMarshaller : IMarshallingGenerator
+    {
+        private static readonly TypeSyntax InteropServicesMarshalType = ParseTypeName(TypeNames.System_Runtime_InteropServices_Marshal);
+        private static readonly TypeSyntax NativeType = PointerType(PredefinedType(Token(SyntaxKind.UShortKeyword)));
+        private static readonly int StackAllocBytesThreshold = 260 * sizeof(ushort);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs:14]: The "logic" for the 260 is based on the `MAX_PATH` variable in Windows. The assumption being that file path are the most used case for strings and therefore we should optimize for the most common length (i.e. 260). I would like to capture that logic so in the future we can make an informed decision about possibly changing it. Feel free to mark this as a `[Compat]`.

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.Interop
+{
+    internal sealed class Utf8StringMarshaller : IMarshallingGenerator
+    {
+        private const int StackAllocBytesThreshold = 260;
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:12]: Same comment here as in UTF-16 marshaller.

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:160]: If that is the plan, then let's call `GetMaxByteCount(1)` during the source gen and use that as the constant.

**Comment by AaronRobinsonMSFT** [comment]
@elinor-fung and @jkoritzinsky From now on when a new `IMarshallingGenerator` is added how about we provide the generated code snippet in the PR? The output from the `T Method(T, in T, ref T, out T)` unit test should be enough to have a conversation.

Thoughts?

File: DllImportGenerator/DllImportGenerator.IntegrationTests/CharacterTests.cs (hunk 5 above)
Comment (elinor-fung) [reply DllImportGenerator/DllImportGenerator.IntegrationTests/CharacterTests.cs:11]: We are now requiring explicit marshalling information for `char` and `string`. I added these to the failure cases in the unit tests.

**Comment by jkoritzinsky** [comment]
That sounds good to me.

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs (hunk 2 above)
Comment (elinor-fung) [reply DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:160]: Bah, so the logic for the calculation is:
- `+ 1` for number of characters in case left over high surrogate is ?
- `* <MaxByteCountPerChar>`
- `+ 1` for null terminator

`GetMaxByteCount` takes into account the +1 for the number of characters, so `GetMaxByteCount(1)` would not give us the constant we want to multiply by.

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs (hunk 4 above)
Comment (elinor-fung) [reply DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs:184]: Yep, that is the plan - was going to do it in a subsequent change

Comment (jkotas) [reply DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf16.cs:184]: Ideally, this would just call https://github.com/dotnet/runtime/issues/25423

File: DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs (hunk 1 above)
Comment (elinor-fung) [reply DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:96]: Good call on no string pinning.

I can keep assigning the stackalloc to a pointer and do:
```
byte* nativePtr;
...
byte* stackAllocPtr = stackalloc byte[byteLen];
Encoding.UTF8.GetBytes(str.AsSpan(), new Span<byte>(stackAllocPtr, byteLen));
nativePtr = stackAllocPtr;
```
Which I think would be equivalent to what you are saying.

Maybe I'm being dense / don't know how to use spans, but if I assign the stackalloc to a `Span`, how do get back at the pointer reasonaably? I know there's `fixed ( ... = &MemoryMarshal.GetReference(...))`, but that seems like extra complication compared to just keeping it as a pointer and constructing the span explicitly?

Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs:96]: Your way with saving the pointer and then constructing the Span would work. I was thinking of using `Unsafe.AsPointer(span[0])` to get the pointer (which since it’s a stackalloc doesn’t need to be pinned). I’m fine if we don’t want to do that though.

**Comment by jkotas** [comment]
These marshalers are not performance critical. It is not important to inline the logic for performance reasons for them. Would it be better to build them using the standard struct-based marshallers that wrap the logic? The marshaling logic can then be optimized independently with ease.

**Comment by AaronRobinsonMSFT** [comment]
>... using the standard struct-based marshallers that wrap the logic?

@jkotas Can you elaborate on this? Are we referencing the internal `StubHelpers` class or is this a well-known pattern?

**Comment by jkotas** [comment]
I meant the `MarshalUsing` attribute from https://github.com/dotnet/runtimelab/blob/DllImportGenerator/DllImportGenerator/designs/StructMarshalling.md.

The idea would be to transform `void F([MarshalAs(UnmanagedType.LPUTF8Str)] string s)` to `void F([MarshalUsing(UTF8StringMarshaler)] string s)` and the rest would just fallout naturally.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas I like this approach for V2 as an extension. For V1, my concern would be where would this and other types live? Pulling this into `NetCoreApp` would require these types eventually live in SPCL, which is something I would like to avoid at all costs. For V2, we could provide an extension library with these types and that is something I am all for.

Does my concern make sense or am I missing something?

**Comment by jkotas** [comment]
The source generator can paste an internal copy of these marshaler types into each library that needs them for now. I think it is better to do that than to manually inline the logic for each use.

I see the inlining of the logic as a important optimization for marshalers that do very little (e.g. just pin). The marshaler type abstraction would be measurably more expensive for those given the current limits of JIT optimizations. The inlining is not worth it for heavy marshalers that allocate memory, copy blocks of memory around, etc. The relative performance overhead should be fairly small for them. In fact, inlining of the more complex logic is likely performance de-optimization overall because of the increased code size. Having the logic for these marshalers in regular C# will make it easy to modify and optimize them.

**Comment by AaronRobinsonMSFT** [comment]
>The source generator can paste an internal copy of these marshaler types into each library that needs them for now. I think it is better to do that than to manually inline the logic for each use.

I was hoping to avoid adding types the user wasn't aware of - not a big deal if we do though. I think this approach is something we should consider post our first pass on NetCoreApp. I could easily be persuaded if we had some performance tests that could help move the needle here and demonstrate the cost/benefit. We are planning on meeting with the performance team today to talk about guidance for performance tests. Once we understand that guidance, performance test authoring was going to start. I will make the `string` scenarios priority one so we can investigate this opportunity. I believe we have time to get in the "everything inlined" approach for some marshallers - I would prefer all - and then optimize as we uncover bottlenecks/opportunities. I do agree there are wins to the approach, but I personally don't understand all the dimensions and would like to have a system to experiment with prior to adding these optimization.

The TL;DR is I want to continue with the current approach and apply it to the `string` marshallers. Once we have performance tests in place we can explore this suggestion fully and measure the tradeoffs. 

@jkotas Is that a fair plan?

**Comment by jkotas** [comment]
The throughput performance is only one of the factors. Other factors to consider include:
- Code/binary size
- Security. Having non-trivial logic operating on unmanaged pointers stamped many times unnecessarily will be frowned upon by the people wearing security hats.

>  demonstrate the cost/benefit

I see the complex marshaling logic factored into a marshaler types as the baseline solution, and the manually inlined code as the extra cost that we need to demonstrate the benefit of.

> I would like to have a system to experiment
> I want to continue with the current approach

Fine with me.

**Review by AaronRobinsonMSFT — APPROVED** [review]
